README
NAME
    Config::IPFilter - Simple, rule-based IP filter

Synopsis
        use Config::IPFilter;
        my $filter = Config::IPFilter->new;
        my $rule   = $filter->add_rule('89.238.128.0', '89.238.191.255', 127,
                                     'Example range');

        # A list of example IPv4 addresses. IPv6 works too.
        my @ipv4 = qw[89.238.156.165 89.238.156.169 89.238.156.170 89.238.167.84
            89.238.167.86 89.238.167.99];

        # Check a list of ips
        say sprintf '%15s is %sbanned', $_, $filter->is_banned($_) ? '' : 'not '
            for @ipv4;

        # Lower the acces level by one pushes it below our ban threshold
        $rule->decrease_access_level;

        # Check a list of ips
        say sprintf '%15s is %sbanned', $_,
            $filter->is_banned($_) ? 'now ' : 'still not '
            for @ipv4;

    You could also load rules directly from an "ipfilter.dat" file.

Description
        # Example of a "ipfilter.dat" file
        #
        # All entered IP ranges will be blocked in both directions. Be careful
        # what you enter here. Wrong entries may totally block access to the
        # network.
        #
        # Format:
        # IP-Range , Access Level , Description
        #
        # Access Levels:
        # 127 blocked
        # >=127 permitted

        064.094.089.000 - 064.094.089.255 , 000 , Gator.com

    This entry will block the IPs from 064.094.089.000 to 064.094.089.255,
    i.e. your code should not connect to any IP in this range.

    At the moment only one, read-only access level is implemented; a value
    at or below 127 means that addresses in that range are banned.

Methods
    Here's a list of 'em...

  my $filter = Config::IPFilter->new( )
    This builds a new, empty object. There are currently no expected
    arguments.

  $filter->add_rule( $rule )
    This method adds a new range to the in-memory ipfilter.

  $filter->add_rule( $lower, $upper, $access_level, $description )
    This method coerces the arguments into a new rule which is then added to
    the in-memory ipfilter.

  $filter->count_rules( )
    Returns a tally of all loaded rules.

  $filter->is_empty( )
    Returns a boolean value indicating whether or not there are any rules
    loaded in the ipfilter.

  $filter->clear_rules( )
    Deletes all rules from the ipfilter.

  $filter->load( $path )
    Slurps an "ipfilter.dat"-like file and adds the rules found inside to
    the ipfilter.

  $filter->save( $path )
    Stores the in-memory ipfilter to disk.

  $filter->is_banned( $ip )
    If $ip is banned, the first rule in which it was found below the
    threshold is returned.

    If not, a false value is returned. Currently, rules with an access_level
    at or below 127 are considered banned.

IPv6 Support
    The standard ipfilter.dat only supports IPv4 addresses but
    Net::BitTorrent's current implementation supports IPv6 as well. Keep
    this in mind when storing an ipfilter.dat file to disk.

Notes
    This is a very good example of code which should not require Moose. In a
    future version, I hope to switch to Moo. ...when "coerce" works to some
    degree.

See Also
    Emule Project's ipfilter.dat documentation
    <http://www.emule-project.net/home/perl/help.cgi?l=1&topic_id=142&rm=sho
    w_topic>

Author
    Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

    CPAN ID: SANKO

License and Legal
    Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

    This program is free software; you can redistribute it and/or modify it
    under the terms of The Artistic License 2.0
    <http://www.perlfoundation.org/artistic_license_2_0>. See the LICENSE
    file included with this distribution or notes on the Artistic License
    2.0 <http://www.perlfoundation.org/artistic_2_0_notes> for
    clarification.

    When separated from the distribution, all original POD documentation is
    covered by the Creative Commons Attribution-Share Alike 3.0 License
    <http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>. See the
    clarification of the CCA-SA3.0
    <http://creativecommons.org/licenses/by-sa/3.0/us/>.



Changes
Version 1.00 | 2010-12-27 17:10:31Z (Mon, 27 Dec 2010) | 9c5350d

    Notes:
     * PAUSE choked on v0.99 so 1.00 brings nothing new

Version 0.99 | 2010-12-27 05:26:21Z (Mon, 27 Dec 2010) | c785a0b

    API Changes/Compatibility Information:
     * See below

    Resolved Issues/Bug Fixes:
     * See below

    Protocol/Behavioral Changes:
     * It exists

    Documentation/Sample Code/Test Suite:
     * See above

    Notes:
     * This code was originally included with Net::BitTorrent as
        Net::BitTorrent::Network::IPFilter


_ -.-. .... .- -. --. . ... _________________________________________
For more information, see the commit log:
    http://github.com/sanko/config-ipfilter/commits/

$Ver$ from git $Rev$
$Date$
$Url$


Build.PL
#!/usr/bin/perl -w
use strict;
use warnings;
use File::Find;
use Module::Build;

$|++;
my $automated_testing = $ENV{'AUTOMATED_TESTING'}
    || $ENV{'PERL_MM_USE_DEFAULT'};
my $is_developer = ((-d '.git') ? 1 : 0) or ($ENV{'RELEASE_TESTING'} ? 1 : 0);
my @tests;
find \&find_cb, qw[t/000_basic/];
my $class = $is_developer ? 'inc::MBX::Developer' : 'Module::Build';

if (!eval "require $class") {
    printf 'Failed to load %s: %s This ain\'t good, so... bye!', $class, $@;
    exit 0;
}
my $mb = $class->new(
     module_name   => 'Config::IPFilter',
     license       => 'artistic_2',
     dist_author   => 'Sanko Robinson <sanko@cpan.org>',
     dist_abstract => 'IPFilter.dat parser',
     requires      => {
                  Moose => 0,
                  perl  => '5.10.0'
     },
     build_requires => {Exporter          => 0,
                        'File::Path'      => 2.07,
                        'File::Spec'      => 0,
                        'File::Temp'      => 0,
                        'Module::Build'   => 0.30,
                        perl              => '5.10.0',
                        'TAP::Harness'    => 3.18,
                        'Test::Class'     => 0.35,
                        'Test::Exception' => 0,
                        'Test::Most'      => 0,
                        ($is_developer
                         ? ('CPAN::Meta::Converter' => 0,
                            'CPAN::Meta::Validator' => 0,
                            JSON                    => 0,
                            'Perl::Tidy'            => 0
                             )
                         : ()
                        )
     },
     configure_requires => {'Module::Build' => 0.36},
     test_files         => \@tests,
     test_file_exts     => ['.t'],
     use_tap_harness    => 1,
     tap_harness_args   => {
                          jobs => $automated_testing ? 9 : 2,
                          rules => {par => ['*']}
     },
     meta_merge => {
         keywords  => [qw[BitTorrent eMule IP Filter IPFilter.dat]],
         resources => {
             bugtracker => {
                      web => 'http://github.com/sanko/config-ipfilter/issues',
                      mailto => 'sanko@cpan.org'
             },
             repository => {
                       url => 'git://github.com/sanko/config-ipfilter.git',
                       web => 'http://github.com/sanko/config-ipfilter/tree/',
                       type => 'git'
             },
             ChangeLog => 'http://github.com/sanko/config-ipfilter/commits/',
             homepage  => 'http://sankorobinson.com/config-ipfilter/',
             license => ['http://www.perlfoundation.org/artistic_license_2_0']
         }
     },
     create_readme => 1
);
$mb->notes(automated_testing => $automated_testing ? 1 : 0);
$mb->notes(release_testing   => $is_developer);
$mb->notes(test_suite        => \@tests);
$mb->notes(gmtime            => gmtime);
$mb->create_build_script;
exit 0;

sub find_cb {
    return if -d $_ or -l $_;
    return unless -T $_;
    return unless $_ =~ m[.+\.t$];
    return push @tests, $File::Find::name;
}

=pod

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

=for rcs $Id$

=cut


examples/synopsis.pl
use strict;
use warnings;
use lib '../lib';
use Config::IPFilter;
use 5.010;

#
my $filter = Config::IPFilter->new;
my $rule =
    $filter->add_rule('89.238.128.0', '89.238.191.255', 127, 'Example range');

# A list of example IPv4 addresses
my @ipv4 = qw[89.238.156.165 89.238.156.169 89.238.156.170 89.238.167.84
    89.238.167.86 89.238.167.99];

# Check a list of ips
say sprintf '%15s is %sbanned', $_, $filter->is_banned($_) ? '' : 'not '
    for @ipv4;

# Lower the acces level by one pushes it below our ban threshold
$rule->decrease_access_level;

# Check a list of ips
say sprintf '%15s is %sbanned', $_,
    $filter->is_banned($_) ? 'now ' : 'still not '
    for @ipv4;

=pod

=head1 Author

=begin :html

L<Sanko Robinson|http://sankorobinson.com/>
<L<sanko@cpan.org|mailto://sanko@cpan.org>> -
L<http://sankorobinson.com/|http://sankorobinson.com/>

CPAN ID: L<SANKO|http://search.cpan.org/~sanko>

=end :html

=begin :text

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=end :text

=head1 License and Legal

Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

=for rcs $Id$

=cut


lib/Config/IPFilter/Rule.pm
package Config::IPFilter::Rule;
{
    use Moose;
    our $MAJOR = 1; our $MINOR = 0; our $DEV = 0; our $VERSION = sprintf('%0d.%02d' . ($DEV ? (($DEV < 0 ? '' : '_') . '%03d') : ('')), $MAJOR, $MINOR, abs $DEV);
    use Config::IPFilter::Types;

    #
    for my $limit (qw[upper lower]) {
        has $limit => (
            isa      => 'Config::IPFilter::Types::Paddr',
            is       => 'ro',
            required => 1,
            coerce   => 1,
            handles  => {
                $limit . '_as_string' => sub {
                    my $s = shift;
                    require Config::IPFilter::Types;
                    Config::IPFilter::Types::paddr2ip($s->{$limit});
                    }
            }
        );
    }
    has description => (isa => 'Str', is => 'ro', required => 1);
    has access_level => (isa      => 'Int',
                         is       => 'ro',
                         required => 1,
                         traits   => ['Counter'],
                         handles  => {
                                     set_access_level      => 'set',
                                     increase_access_level => 'inc',
                                     decrease_access_level => 'dec'
                         }
    );

    sub in_range {
        my ($s, $ip) = @_;
        $ip = Config::IPFilter::Types::ip2paddr($ip);
        return (($s->lower lt $ip && $s->upper gt $ip) ? 1 : 0);
    }

    sub _as_string {
        my $s = shift;
        return sprintf '%s - %s, %d, %s', $s->lower_as_string,
            $s->upper_as_string,
            $s->access_level, $s->description;
    }
    no Moose;
}
1;

=pod

=head1 NAME

Config::IPFilter::Rule - A single block of IP addresses

=head1 Synopsis

    use Config::IPFilter::Rule;
    my $rule =
        Config::IPFilter::Rule->new(access_level => 255,
                                    lower        => '0.0.0.0',
                                    upper        => '255.255.255.255',
                                    description  => 'teh innernetz'
        );

    # Hide in a cave
    $rule->set_access_level(126);

IPv6 is also supported...

    use Config::IPFilter::Rule;
    my $rule = Config::IPFilter::Rule->new(
           access_level => 255,
           upper        => 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff',
           lower        => '::',  # ojai
           description  => 'teh innernetz'
    );

    # Hide in a cave
    $rule->set_access_level(126);
    $filter->is_banned('2001:db8::');    # It, and everything else, is banned

=head1 Description

This is a single range of addresses (IPv4 or IPv6) which all share a single
access level.

=head1 my $rule = Config::IPFilter::Rule->B<new>( ... )

This constructs a new object. The following arguments are required:

=over

=item C<access_level>

This is an integer value.

=item C<description>

This is a string. You should put the reason why this range exists here.

=item C<upper>

This is the address at the highest end of this range.

=item C<lower>

This is the address at the lowest end of this range.

=back

=head1 $filter->B<access_level>( )

Returns the access level currently defined for this range. See
L<< Config::IPFilter->is_banned( ... )|Config::IPFilter/"$filter->B<is_banned>( $ip )" >>.

=head1 $filter->B<set_access_level>( $value )

Sets the access level for this range.

=head1 $filter->B<increase_access_level>( [ $inc ] )

Sets the access level for this range C<$inc> degrees higher. The default value
of C<$inc> is C<1>.

=head1 $filter->B<decrease_access_level>( [ $dec ] )

Sets the access level for this range C<$dec> degrees lower. The default value
of C<$dec> is C<1>.

=head1 $filter->B<in_range>( $address )

If the given address is within this rule's range, a true value is returned
otherwise a false value is returned.

=head1 $filter->B<description>( )

This is original string value you passed during construction.

=head1 $filter->B<upper>( )

This returns the address at the highest end of the range.

=head1 $filter->B<lower>( )

This returns the address at the lowest end of the range.

=head1 Author

=begin :html

L<Sanko Robinson|http://sankorobinson.com/>
<L<sanko@cpan.org|mailto://sanko@cpan.org>> -
L<http://sankorobinson.com/|http://sankorobinson.com/>

CPAN ID: L<SANKO|http://search.cpan.org/~sanko>

=end :html

=begin :text

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=end :text

=head1 License and Legal

Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=for rcs $Id$

=cut


lib/Config/IPFilter/Types.pm
package Config::IPFilter::Types;
{
    use Moose::Util::TypeConstraints;
    our $MAJOR = 1; our $MINOR = 0; our $DEV = 0; our $VERSION = sprintf('%0d.%02d' . ($DEV ? (($DEV < 0 ? '' : '_') . '%03d') : ('')), $MAJOR, $MINOR, abs $DEV);

    # IPv6 packed address
    subtype 'Config::IPFilter::Types::Paddr' => as 'Str' =>
        where { length $_ == 16 } =>
        message { sprintf '%s is not 16 bytes', $_ };
    coerce 'Config::IPFilter::Types::Paddr' => from 'Str' =>
        via { ip2paddr($_) };

    #
    subtype 'Config::IPFilter::Types::Addr' => as 'ArrayRef' =>
        where { $#{$_[0]} == 1 }   => message {'looking for [host, port]'} =>
        where { defined $_[0][0] } => message {'hostname is missing'} =>
        where { defined $_[0][1] } => message {'port is missing'} =>
        where { $_[0][1] =~ m[^\d+$] } => message {'malformed port'};

    # Utility function
    sub ip2paddr {
        my $addr = shift;
        $addr = '::' . $addr unless $addr =~ /:/;
        if ($addr =~ /^(.*:)(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/)
        {    # mixed hex, dot-quad
            return undef if $2 > 255 || $3 > 255 || $4 > 255 || $5 > 255;
            $addr = sprintf('%s%X%02X:%X%02X', $1, $2, $3, $4, $5)
                ;    # convert to pure hex
        }
        my $c;
        return undef
            if $addr =~ /[^:\da-f]/i ||    # non-hex character
                (($c = $addr) =~ s/::/x/ && $c =~ /(?:x|:):/)
                ||                         # double :: ::?
                $addr =~ /[0-9a-fA-F]{5,}/;    # more than 4 digits
        $c = $addr =~ tr[:][:];                # count the colons
        return undef if $c < 7 && $addr !~ /::/;
        if ($c > 7) {                          # strip leading or trailing ::
            return undef
                unless $addr =~ s|^::|:|
                    || $addr =~ s|::$|:|;
            return undef if --$c > 7;
        }
        $addr =~ s|::|:::| while $c++ < 7;     # expand compressed fields
        $addr .= 0 if $addr =~ m[:$];
        my @hex = split ':', $addr;
        $hex[$_] = hex $hex[$_] || 0 for 0 .. $#hex;
        return pack 'n8', @hex;
    }

    sub paddr2ip ($) {
        return inet_ntoa($_[0]) if length $_[0] == 4;    # ipv4
        return inet_ntoa($1)
            if length $_[0] == 16
                && $_[0] =~ m[^\0{10}\xff{2}(.{4})$];    # ipv4
        return unless length($_[0]) == 16;
        my @hex = (unpack('n8', $_[0]));
        $hex[9] = $hex[7] & 0xff;
        $hex[8] = $hex[7] >> 8;
        $hex[7] = $hex[6] & 0xff;
        $hex[6] >>= 8;
        my $return = sprintf '%X:%X:%X:%X:%X:%X:%D:%D:%D:%D', @hex;
        $return =~ s|(0+:)+|:|x;
        $return =~ s|^0+    ||x;
        $return =~ s|^:+    |::|x;
        $return =~ s|::0+   |::|x;
        $return =~ s|^::(\d+):(\d+):(\d+):(\d+)|$1.$2.$3.$4|x;
        return $return;
    }

    #
    no Moose::Util::TypeConstraints;
}
1;

=pod

=head1 NAME

Config::IPFilter::Types - Moose Types for Config::IPFilter

=head1 Description

Nothing to see here (yet), folks. Keep it movin'.

=head1 Author

=begin :html

L<Sanko Robinson|http://sankorobinson.com/>
<L<sanko@cpan.org|mailto://sanko@cpan.org>> -
L<http://sankorobinson.com/|http://sankorobinson.com/>

CPAN ID: L<SANKO|http://search.cpan.org/~sanko>

=end :html

=begin :text

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=end :text

=head1 License and Legal

Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

Neither this module nor the L<Author|/Author> is affiliated with BitTorrent,
Inc.

=for rcs $Id$

=cut


lib/Config/IPFilter.pm
package Config::IPFilter;
{
    use Moose;
    our $MAJOR = 1; our $MINOR = 00; our $DEV = 0; our $VERSION = sprintf('%0d.%02d' . ($DEV ? (($DEV < 0 ? '' : '_') . '%03d') : ('')), $MAJOR, $MINOR, abs $DEV);
    use Config::IPFilter::Types;

    #
    has rules => (isa      => 'ArrayRef[Config::IPFilter::Rule]',
                  is       => 'ro',
                  traits   => ['Array'],
                  init_arg => undef, # XXX - In the future, allow this
                  default  => sub { [] },
                  handles  => {
                              add_rule            => 'push',
                              count_rules         => 'count',
                              is_empty            => 'is_empty',
                              get_rule            => 'get',
                              first_rule          => 'first',
                              grep_rules          => 'grep',
                              map_rules           => 'map',
                              sort_rules          => 'sort',
                              sort_rules_in_place => 'sort_in_place',
                              shuffle_rules       => 'shuffle',
                              clear_rules         => 'clear',
                              insert_rule         => 'insert',
                              delete_rule         => 'delete',
                              push_rule           => 'push',
                              pop_rule            => 'pop'
                  }
    );
    around add_rule => sub {
        my ($c, $s, $l, $u, $a, $d) = @_;
        $l = blessed $l? $l : sub {
            require Config::IPFilter::Rule;
            Config::IPFilter::Rule->new(lower        => $l,
                                        upper        => $u,
                                        access_level => $a,
                                        description  => $d
            );
            }
            ->();
        return $c->($s, $l) ? $l : ();
    };

    sub load {
        my ($s, $path) = @_;
        open(my $IPFilter, '<', $path) || return;
        for my $line (<$IPFilter>) {
            next if $line =~ m[(?:^#|^$)];
            my ($range, $access_level, $desc) =
                ($line =~ m[^(.+-.+)\s*,\s*(\d+)\s*,\s*(.+)\s*$]);
            next if !$range;
            my ($start, $end) = ($range =~ m[^(.+)\s*-\s*(.+)\s*$]);
            $_ =~ s[\s][]g for $start, $end;
            $s->add_rule($start, $end, $access_level, $desc);
        }
        1;
    }

    sub save {
        my ($s, $path) = @_;
        open(my $IPFilter, '>', $path) || return;
        for my $rule (
            $s->sort_rules(
                sub {
                    $_[0]->lower cmp $_[1]->lower
                        || $_[0]->upper cmp $_[1]->upper;
                }
            )
            )
        {   syswrite $IPFilter, $rule->_as_string . "\n";
        }
        return close $IPFilter;
    }

    sub is_banned {
        my ($s, $ip) = @_;
        return $s->first_rule(
            sub {
                $_->in_range($ip) && $_->access_level < 127;
            }
        ) || ();
    }

    #
    no Moose;
}
1;

=pod

=head1 NAME

Config::IPFilter - Simple, rule-based IP filter

=head1 Synopsis

    use Config::IPFilter;
    my $filter = Config::IPFilter->new;
    my $rule   = $filter->add_rule('89.238.128.0', '89.238.191.255', 127,
                                 'Example range');

    # A list of example IPv4 addresses. IPv6 works too.
    my @ipv4 = qw[89.238.156.165 89.238.156.169 89.238.156.170 89.238.167.84
        89.238.167.86 89.238.167.99];

    # Check a list of ips
    say sprintf '%15s is %sbanned', $_, $filter->is_banned($_) ? '' : 'not '
        for @ipv4;

    # Lower the acces level by one pushes it below our ban threshold
    $rule->decrease_access_level;

    # Check a list of ips
    say sprintf '%15s is %sbanned', $_,
        $filter->is_banned($_) ? 'now ' : 'still not '
        for @ipv4;

You could also load rules directly from an C<ipfilter.dat> file.

=head1 Description

    # Example of a "ipfilter.dat" file
    #
    # All entered IP ranges will be blocked in both directions. Be careful
    # what you enter here. Wrong entries may totally block access to the
    # network.
    #
    # Format:
    # IP-Range , Access Level , Description
    #
    # Access Levels:
    # 127 blocked
    # >=127 permitted

    064.094.089.000 - 064.094.089.255 , 000 , Gator.com

This entry will block the IPs from 064.094.089.000 to 064.094.089.255, i.e.
your code should not connect to any IP in this range.

At the moment only one, read-only access level is implemented; a value at or
below C<127> means that addresses in that range are banned.

=head1 Methods

Here's a list of 'em...

=head2 my $filter = Config::IPFilter->B<new>( )

This builds a new, empty object. There are currently no expected arguments.

=head2 $filter->B<add_rule>( $rule )

This method adds a new L<range|Config::IPFilter::Rule> to the in-memory
ipfilter.

=head2 $filter->B<add_rule>( $lower, $upper, $access_level, $description )

This method coerces the arguments into a new L<rule|Config::IPFilter::Rule>
which is then added to the in-memory ipfilter.

=head2 $filter->B<count_rules>( )

Returns a tally of all loaded L<rule|Config::IPFilter::Rule>s.

=head2 $filter->B<is_empty>( )

Returns a boolean value indicating whether or not there are any
L<rule|Config::IPFilter::Rule>s loaded in the ipfilter.

=head2 $filter->B<clear_rules>( )

Deletes all L<rule|Config::IPFilter::Rule>s from the ipfilter.

=head2 $filter->B<load>( $path )

Slurps an C<ipfilter.dat>-like file and adds the
L<rule|Config::IPFilter::Rule>s found inside to the ipfilter.

=head2 $filter->B<save>( $path )

Stores the in-memory ipfilter to disk.

=head2 $filter->B<is_banned>( $ip )

If C<$ip> is banned, the first L<rule|Config::IPFilter::Rule> in which it was
found below the threshold is returned.

If not, a false value is returned. Currently, rules with an
L<< access_level|Config::IPFilter::Rule/"$filter->B<access_level>( )" >> at or
below C<127> are considered banned.

=head1 IPv6 Support

The standard ipfilter.dat only supports IPv4 addresses but
L<Net::BitTorrent>'s current implementation supports IPv6 as well. Keep this
in mind when L<storing|/save> an ipfilter.dat file to disk.

=head1 Notes

This is a very good example of code which should not require L<Moose|Moose>.
In a future version, I hope to switch to L<Moo|Moo>. ...when C<coerce> works
to some degree.

=head1 See Also

L<Emule Project's ipfilter.dat documentation|http://www.emule-project.net/home/perl/help.cgi?l=1&topic_id=142&rm=show_topic>

=head1 Author

=begin :html

L<Sanko Robinson|http://sankorobinson.com/>
<L<sanko@cpan.org|mailto://sanko@cpan.org>> -
L<http://sankorobinson.com/|http://sankorobinson.com/>

CPAN ID: L<SANKO|http://search.cpan.org/~sanko>

=end :html

=begin :text

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=end :text

=head1 License and Legal

Copyright (C) 2010, 2011 by Sanko Robinson <sanko@cpan.org>

This program is free software; you can redistribute it and/or modify it under
the terms of
L<The Artistic License 2.0|http://www.perlfoundation.org/artistic_license_2_0>.
See the F<LICENSE> file included with this distribution or
L<notes on the Artistic License 2.0|http://www.perlfoundation.org/artistic_2_0_notes>
for clarification.

When separated from the distribution, all original POD documentation is
covered by the
L<Creative Commons Attribution-Share Alike 3.0 License|http://creativecommons.org/licenses/by-sa/3.0/us/legalcode>.
See the
L<clarification of the CCA-SA3.0|http://creativecommons.org/licenses/by-sa/3.0/us/>.

=for rcs $Id$

=cut
